index/miner/lotusidx/store/store.go

```diff
@@ -0,0 +1,189 @@
+package store
+
+import (
+	"context"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+
+	"github.com/ipfs/go-datastore"
+	kt "github.com/ipfs/go-datastore/keytransform"
+	"github.com/ipfs/go-datastore/query"
+	logger "github.com/ipfs/go-log/v2"
+	"github.com/textileio/powergate/v2/index/miner"
+)
+
+var (
+	dsMetadata      = datastore.NewKey("meta")
+	dsOnChainMiner  = datastore.NewKey("onchain/miner")
+	dsOnChainHeight = datastore.NewKey("onchain/height")
+
+	log = logger.Logger("lotusidx-store")
+)
+
+// Store is a store to save on-chain and metadata information about the chain.
+type Store struct {
+	ds *kt.Datastore
+}
+
+// New returns a new *Store.
+func New(ds *kt.Datastore) (*Store, error) {
+	return &Store{
+		ds: ds,
+	}, nil
+}
+
+// SaveMetadata creates/updates metadata information of miners.
+func (s *Store) SaveMetadata(index miner.MetaIndex) error {
+	for addr, meta := range index.Info {
+		key := makeMinerMetadataKey(addr)
+		buf, err := json.Marshal(meta)
+		if err != nil {
+			return fmt.Errorf("marshaling metadata for miner %s: %s", addr, err)
+		}
+		if err := s.ds.Put(key, buf); err != nil {
+			return fmt.Errorf("saving metadata in store: %s", err)
+		}
+	}
+	return nil
+}
+
+// SaveOnChain creates/updates on-chain information of miners.
+func (s *Store) SaveOnChain(ctx context.Context, index miner.ChainIndex) error {
```

Comment:
The miner index has a lot of miners. I used a `datastore.Batching` interface to allow to persist things in batches since single updates were too slow.